modules/statistics/AnalyticsAdapter.js

```diff
@@ -1,130 +1,297 @@
+
+        if (!event.action || !event.actionSubject || !event.source) {
+            logger.error(
+                'Required field missing (action, actionSubject or source)');
+
+            return false;
+        }
+
+        // Track events have additional required fields.
+        if (type === TYPE_TRACK) {
+            event.objectType = event.objectType || 'generic-object-type';
+            event.containerType
+                = event.containerType || 'generic-container-type';
+
+            if (!event.objectType || !event.objectId
+                || !event.containerType || !event.containerId) {
+                logger.error(
+                    'Required field missing (containerId, containerType, '
+                        + 'objectId or objectType)');
+
+                return false;
+            }
+        }
+
+        return true;
     }
 
     /**
-     * Dispose analytics. Clears all handlers.
+     * Saves an event to the cache, if the cache is enabled.
+     * @param event the event to save.
+     * @returns {boolean} true if the event was saved, and false otherwise (i.e.
+     * if the cache was disabled).
+     * @private
      */
-    dispose() {
-        cacheAnalytics.drainCachedEvents();
-        this.analyticsHandlers.clear();
-        this.disposed = true;
+    _maybeCacheEvent(event) {
+        if (this.cache) {
+            this.cache.push(event);
+
+            // We limit the size of the cache, in case the user fails to ever
+            // set the analytics handlers.
+            if (this.cache.length > MAX_CACHE_SIZE) {
+                this.cache.splice(0, 1);
+            }
+
+            return true;
+        }
+
+        return false;
+
     }
 
     /**
-     * Sets the handlers that are going to be used to send analytics and send
-     * the cached events.
-     * @param {Array} handlers the handlers
+     *
+     * @param event
+     * @private
      */
-    setAnalyticsHandlers(handlers) {
-        if (this.disposed) {
-            return;
+    _sendEvent(event) {
+        if (this._maybeCacheEvent(event)) {
+            // The event was consumed by the cache.
+        } else {
+            // We append the permanent properties at the time we send the event,
+            // not at the time we receive it.
+            const extendedEvent = this._appendPermanentProperties(event);
+
+            this.analyticsHandlers.forEach(
+                handler => handler.sendEvent(extendedEvent));
         }
-        this.analyticsHandlers = new Set(handlers);
-        cacheAnalytics.drainCachedEvents().forEach(
-            ev => this.sendEvent(ev.eventName, ev.data));
     }
 
     /**
-     * Adds map of properties that will be added to every event.
-     * @param {Object} properties the map of properties
+     * Extends an event object with the configured permanent properties.
+     * @param event the event to extend with permanent properties.
+     * @returns {any & ({}|*)} the extended event
```

Comment:
What is this return type? Also, Object.assign will modify in place. Just fyi.

Reply:
Yes, that's the intention. Except I wanted to assign to event.attributes. Fixed now.